mcs/build/gensources.cs

```diff
@@ -0,0 +1,411 @@
+    public string Directory;
+    public string Pattern;
+    public string HostPlatform;
+    public string ProfileName;
+}
+
+public struct Source {
+    public string FileName;
+}
+
+public class ParseResult {
+    public readonly string LibraryDirectory, LibraryName;
+
+    public readonly List<ParseEntry> Sources = new List<ParseEntry> ();
+    public readonly List<ParseEntry> Exclusions = new List<ParseEntry> ();
+
+    // FIXME: This is a bad spot for this value but enumerators don't have outparam support
+    public int ErrorCount = 0;
+
+    public ParseResult (string libraryDirectory, string libraryName) {
+        LibraryDirectory = libraryDirectory;
+        LibraryName = libraryName;
+    }
+
+    private static string GetRelativePath (string fullPath, string relativeToDirectory) {
+        fullPath = fullPath.Replace (SourcesParser.DirectorySeparator, "/");
+        relativeToDirectory = relativeToDirectory.Replace (SourcesParser.DirectorySeparator, "/");
+
+        if (!relativeToDirectory.EndsWith (SourcesParser.DirectorySeparator))
+            relativeToDirectory += SourcesParser.DirectorySeparator;
+        var dirUri = new Uri (relativeToDirectory);
+        var pathUri = new Uri (fullPath);
+
+        var relativeUri = Uri.UnescapeDataString (
+            dirUri.MakeRelativeUri (pathUri).OriginalString
+        ).Replace ("/", SourcesParser.DirectorySeparator);
+
+        if (SourcesParser.TraceLevel >= 4)
+            Console.Error.WriteLine ($"// {fullPath} -> {relativeUri}");
+
+        return relativeUri;
+    }
+
+    private IEnumerable<string> EnumerateMatches (
+        IEnumerable<ParseEntry> entries,
+        string hostPlatformName, string profileName
+    ) {
+        foreach (var entry in entries) {
+            if (
+                (hostPlatformName != null) &&
+                (entry.HostPlatform ?? hostPlatformName) != hostPlatformName
+            )
+                continue;
+            if (
+                (profileName != null) &&
+                (entry.ProfileName ?? profileName) != profileName
+            )
+                continue;
+
+            var absolutePath = Path.Combine (entry.Directory, entry.Pattern);
+            var absoluteDirectory = Path.GetDirectoryName (absolutePath);
+            var absolutePattern = Path.GetFileName (absolutePath);
+
+            if (SourcesParser.TraceLevel >= 3) {
+                if ((absolutePattern != entry.Pattern) || (absoluteDirectory != entry.Directory))
+                    Console.Error.WriteLine ($"// {entry.Directory} / {entry.Pattern} -> {absoluteDirectory} / {absolutePattern}");
+            }            
+
+            if (!Directory.Exists (absoluteDirectory)) {
+                Console.Error.WriteLine ($"Directory does not exist: {Path.GetFullPath (absoluteDirectory)}");
+                ErrorCount += 1;
+                continue;
+            }
+
+            var matchingFiles = Directory.GetFiles (absoluteDirectory, absolutePattern);
```

Comment:
We should only call `.GetFiles()` when we actually have a `*` pattern, it is quite expensive to do it for each file otherwise. This would save 100ms when generating net_4_x corlib.